File: LeetCode75/20_Valid_Parentheses/20_Valid_Parentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        val bracketPairs = mapOf(
            ')' to '(',
            '}' to '{',
            ']' to '['
        )

        // スタックを使って括弧を管理
        val stack = mutableListOf<Char>()

        for (char in s) {
            if (char in bracketPairs.values) {
                // 開き括弧の場合、stackに追加
                stack.add(char)
            } else if (char in bracketPairs.keys) {
                // 閉じ括弧の場合、対応する開き括弧と比較
                if (stack.isEmpty() || stack.removeAt(stack.size - 1) != bracketPairs[char]) {
                    return false // 対応しない場合は無効
                }
            }
        }
        // スタックが空なら全て有効
        return stack.isEmpty()
    }
}


/*
【アプローチ】
・()と[]と{}これがtrue
 */

File: LeetCode75/20_Valid_Parentheses/20_Valid_Parentheses.ts
function isValid(s: string): boolean {
    const map = new Map<string, string> ([
        [')', '('],
        ['}', '{'],
        [']', '['],
    ]);
    const stack: string[] = [];

    for (let char of s) {
        if (map.has(char)) {
            const topElement = stack.length > 0 ? stack.pop() : "#";

            if (map.get(char) !== topElement) {
                return false;
            }
        } else {
            stack.push(char);
        }
    }

    return stack.length === 0;
}

/**
 【実装】
 1.閉じ括弧と開き括弧のkeyとvalueを持つ変数を宣言
 2.与えられてきた開き括弧の文字列の配列を宣言する
 3.sの配列の要素をループ処理
 4.もし、map変数にキーが存在する場合
    - stackの最後の値を取得
    - もし、mapのキー値と等しくない場合
        - falseを返す
 5.もし、map変数にキーが存在しない場合
 */
